azure/services/virtualmachineimages/images.go

```diff
@@ -0,0 +1,192 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package virtualmachineimages
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/blang/semver"
+	"github.com/pkg/errors"
+	infrav1 "sigs.k8s.io/cluster-api-provider-azure/api/v1beta1"
+	"sigs.k8s.io/cluster-api-provider-azure/azure"
+	"sigs.k8s.io/cluster-api-provider-azure/util/tele"
+)
+
+// Service provides operations on Azure VM Images.
+type Service struct {
+	Client
+	azure.Authorizer
+}
+
+// New creates a new VM Images service.
+func New(auth azure.Authorizer) *Service {
+	return &Service{
+		Client:     NewClient(auth),
+		Authorizer: auth,
+	}
+}
+
+// GetDefaultUbuntuImage returns the default image spec for Ubuntu.
+func (s *Service) GetDefaultUbuntuImage(ctx context.Context, location, k8sVersion string) (*infrav1.Image, error) {
+	v, err := semver.ParseTolerant(k8sVersion)
+	if err != nil {
+		return nil, errors.Wrapf(err, "unable to parse Kubernetes version \"%s\"", k8sVersion)
+	}
+
+	osVersion := getUbuntuOSVersion(v.Major, v.Minor, v.Patch)
+	publisher, offer := azure.DefaultImagePublisherID, azure.DefaultImageOfferID
+	skuID, version, err := s.getDefaultImageSKUIDAndVersion(
+		ctx, location, publisher, offer, k8sVersion, fmt.Sprintf("ubuntu-%s", osVersion))
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get default image")
+	}
+
+	defaultImage := &infrav1.Image{
+		Marketplace: &infrav1.AzureMarketplaceImage{
+			Publisher: publisher,
+			Offer:     offer,
+			SKU:       skuID,
+			Version:   version,
+		},
+	}
+
+	return defaultImage, nil
+}
+
+// GetDefaultWindowsImage returns the default image spec for Windows.
+func (s *Service) GetDefaultWindowsImage(ctx context.Context, location, k8sVersion, runtime, osAndVersion string) (*infrav1.Image, error) {
+	v122 := semver.MustParse("1.22.0")
+	v, err := semver.ParseTolerant(k8sVersion)
+	if err != nil {
+		return nil, errors.Wrapf(err, "unable to parse Kubernetes version \"%s\"", k8sVersion)
+	}
+
+	// If containerd is specified we don't currently support less than 1.22
+	if v.LE(v122) && runtime == "containerd" {
+		return nil, errors.New("containerd image only supported in 1.22+")
+	}
+
+	if osAndVersion == "" {
+		osAndVersion = azure.DefaultWindowsOsAndVersion
+	}
+
+	publisher, offer := azure.DefaultImagePublisherID, azure.DefaultWindowsImageOfferID
+	skuID, version, err := s.getDefaultImageSKUIDAndVersion(
+		ctx, location, publisher, offer, k8sVersion, osAndVersion)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get default image")
+	}
+
+	// Starting with 1.22 we default to containerd for Windows unless the runtime flag is set.
+	if v.GTE(v122) && runtime != "dockershim" {
```

Comment:
given #2292 merged, can we remove this dockershim image finding logic in a follow up @jsturtevant?